src/query/Index.hpp

```diff
@@ -74,22 +75,27 @@ class Index {
   std::vector<VertexID> getVerticesInsideBox(const mesh::BoundingBox &bb);
 
   /**
-   * @brief Find the closest interpolation element to the given location. 
+   * @brief Find the closest interpolation element to the given location.
    * If exists, triangle or edge projection element is returned. If not vertex projection element, which is the nearest neighbor is returned.
-   * 
-   * param[in] sourceVertex 
+   *
+   * param[in] sourceVertex
    * param[in] n how many nearest edges/faces are going to be checked
-   * 
+   *
    * param[out] pair of interpolation and the distance to corresponding vertex/edge/triangle
    *
-  */
+   */
   ProjectionMatch findNearestProjection(const Eigen::VectorXd &location, int n);
 
+  /// Clear the index
+  void clear();
+
 private:
-  struct IndexImpl;
+  class IndexImpl;
   std::unique_ptr<IndexImpl> _pimpl;
 
-  const mesh::PtrMesh             _mesh;
+  /// The indexed Mesh.
+  mesh::Mesh *_mesh;
```

Comment:
Why use raw pointer instead of smart pointers?

Reply:
Isn't that the purpose of weak pointers?

Reply:
Weak pointers model a non-owning handle to a shared pointer.
To access the object, you tell the weak pointer that you require ownership, then it returns an owning shared pointer. Think of deferred ownership.